src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp

```diff
@@ -724,6 +725,9 @@ size_t ShenandoahHeapRegion::setup_sizes(size_t max_heap_size) {
     FLAG_SET_DEFAULT(ShenandoahMinRegionSize, MIN_REGION_SIZE);
   }
 
+  // Generational Shenandoah needs this alignment for card tables.
```

Comment:
Thank you for this fix! It would be nice if this constraint were only applied for generation mode, but these sizes are computed quite earlier during startup. You'd need to factor the code out of `ShenandoahHeap::initialize_heuristics` to know whether the constraint is required at this point.

Reply:
Yes, I thought I'd start with a simple fix to highlight the problem first. I experimented with exactly what you suggested, parsing ShenandoahGCMode. I can update this the 2MB alignment isn't desired unconditionally.

Reply:
Yes please. We've tried to limit the impact of the generational mode on Shenandoah's other modes.